Add more docs around migration from RStudio (#116)
posit-dev/positron-website: migrate-rstudio-code.qmd

---
title: "Writing R code"
aliases:
  - r-snippets.qmd
---

Writing R code in Positron should feel quite familiar to RStudio users.
However two aspects of the workflow might feel a bit different:

* Formatting via Air
* Snippets, both built-in and custom

## Formatting R code with Air

Positron can format R code with [Air](https://posit-dev.github.io/air/), a new R formatter whose creation was largely motivated by Positron.

> A formatter takes charge of the layout of your R code, which refers to whitespace, newlines, indentation, and the like.

Author: I feel like this is sort of crying out for an image that shows before Air formatting vs. after for a bit of code.

The goal is to enforce a set of style conventions that enhance readability and robustness.
Air shares properties with other modern formatters and linters, such as Black and Ruff for Python:

* It's highly opinionated and offers relatively few options for customization.
* It's written in Rust, which makes it extremely fast.

> ### How does Positron get Air?

Reviewer: Maybe this is `How do I install Air?` or something like that, to make it more user centric

Air is fundamentally a command line tool, but fortunately the Air CLI ships with Positron.
Most Positron users do not need to do anything intentional to install Air or to keep it up-to-date.

> Under the hood, Air is packaged in a VS Code extension and that extension is one of the so-called [bootstrapped extensions](extensions.qmd) that come with a fresh Positron installation.

Reviewer 2: ```suggestion
Under the hood, Air is packaged in a VS Code extension and that extension is one of the so-called [bootstrapped extensions](extensions.qmd#bootstrapped-extensions) that are installed for you in Positron.
```

The Air extension is published on [Open VSX](https://open-vsx.org/extension/posit/air-vscode), the marketplace used by IDEs like Positron and Cursor, and also on the [Visual Studio Code Marketplace](https://marketplace.visualstudio.com/items?itemName=Posit.air-vscode), which serves only VS Code.

### When and how does Air run?

There are various explicit and implicit gestures to run Air over your code:

* "Format on save": formats R code every time you save a file. This is a lifestyle you must opt in to, but it is our strong recommendation that you do so. Below we give details on how to set this up.
* Formatting commands: These are explicit gestures to format specific bits of code.
  - *Format Selection*
  - *Format Document*
  - *Format Cell*
  - *Air: Format Workspace Folder*

  These are all available in the command palette.

> R code receives some formatting as you type, but these rules are based on regular expressions, which fundamentally limits their capability.

Author: Interesting. I'm not intending to promise some specific setting ("Format on type"), but rather a more colloquial "yeah, sure, some basic formatting is happening while you type". I'm just trying to explain why the code had some semblance of formatting to begin with, then Air comes through and does even more. Would appreciate wording suggestions, if we think this what I've got is misleading.

A second formatting pass with Air, when you save the file or intentionally format it, produces a much better result, since Air works off of an actual syntax tree.

### Opt in to "Format on save"

> The easiest way to get our recommended set up is to run [`usethis::use_air()`](https://usethis.r-lib.org/dev/reference/use_air.html) (note that this currently requires the development version of usethis).

Author: I am going to work on a usethis release in the very near future.

Reviewer 3: I also think given some of the opinionated stance of air, we should have a callout to skip for small areas of code where the user "disagrees" with our formatting choices: https://posit-dev.github.io/air/formatter.html#skip-comments
